sw/cheri/common/sdcard-utils.hh

```diff
@@ -0,0 +1,472 @@
+  static constexpr uint8_t ones[] = {0xffu, 0xffu, 0xffu, 0xffu, 0xffu, 0xffu, 0xffu, 0xffu, 0xffu, 0xffu};
+
+ public:
+  // Transfers in SPI mode are always in terms of 512-byte blocks.
+  static constexpr unsigned kBlockLen = 512u;
+
+  // Supplied with SPI controller and its appropriate CS line, and a GPIO block with its input pin
+  // number. The SPI controller performs the communication with the SD card and the GPIO simply
+  // provides the card detection indicator (low = present).
+  //
+  // CRC calculation is turned on by default, although this requires the CPU to perform a little
+  // bit of additional work. It may be disabled if desired. SPI mode communication may be performed
+  // with or without CRC checking.
+  //
+  // Logging may optionally be requested.
+  SdCard(volatile SonataSpi *spi_, volatile SonataGpioBase *gpio_, unsigned cs_ = 1u, unsigned det_ = 16u,
+         bool crc_ = true, Log *log_ = nullptr)
+      : spi(spi_), gpio(gpio_), cs(1u << cs_), det(1u << det_), crcOn(crc_), log(log_) {}
+
+  // SD command codes. (Section 7.3.1)
+  enum CmdCode {
+    CMD_GO_IDLE_STATE        = 0,
+    CMD_SEND_OP_COND         = 1,
+    CMD_SEND_IF_COND         = 8,
+    CMD_SEND_CSD             = 9,
+    CMD_SEND_CID             = 10,
+    CMD_STOP_TRANSMISSION    = 12,
+    CMD_SET_BLOCKLEN         = 16,
+    CMD_READ_SINGLE_BLOCK    = 17,
+    CMD_READ_MULTIPLE_BLOCK  = 18,
+    CMD_WRITE_SINGLE_BLOCK   = 24,
+    CMD_WRITE_MULTIPLE_BLOCK = 25,
+    SD_SEND_OP_COND          = 41,
+    CMD_APP_CMD              = 55,
+    CMD_READ_OCR             = 58,
+    CMD_CRC_ON_OFF           = 59,
+  };
+
+  // SD Control Tokens. (Section 7.3.3)
+  enum CtrlToken {
+    // Start Block Token precedes data block, for all but Multiple Block Write.
+    StartBlockToken = 0xfeu,
+    // Start Block Token used for Multiple Block Write operations.
+    StartBlockTokenMW = 0xfcu,
+    // Stop Transaction Token, for Multiple Block Writes.
+    StopTranToken = 0xfdu,
+  };
+
+  // Indicates whether there is an SD card present in the slot.
+  bool present() const { return !(gpio->debouncedInput & det); }
+
+  void select_card(bool enable) { spi->cs = enable ? (spi->cs & ~cs) : (spi->cs | cs); }
+
+  // Initialise the SD card ready for use.
+  bool init() {
+    // Every card tried seems to be more than capable of keeping up with 20Mbps.
+    constexpr unsigned kSpiSpeed = 0u;
+    spi->init(false, false, true, kSpiSpeed);
+
+    // Apparently we're required to send at least 74 SD CLK cycles with
+    // the device _not_ selected before talking to it.
+    spi->blocking_write(ones, 10);
+    spi->wait_idle();
+
+    select_card(true);
+
+    // Note that this is a very stripped-down card initialisation sequence
+    // that assumes SDHC version 2, so use a more recent microSD card.
+    do {
+      send_command(CMD_GO_IDLE_STATE, 0u);
+    } while (0x01 != get_response_R1());
+
+    send_command(CMD_SEND_IF_COND, 0x1aau);
```

Comment:
Supporting a v1 card is not actually hard: if the card does not answer this command then it's v1. Maybe at least print an error message and bail out?